File: solutions/fortran/sieve/sieve.f90
module sieve
   implicit none
contains

   pure function primes(limit)
      integer, intent(in) :: limit
      integer, allocatable :: primes(:)
      integer :: i, j
      logical :: marked(2:limit)
      marked = .false.

      do i = 2,floor(sqrt(real(limit)))
         if ( .not. marked(i) ) then
            do j = i*i,limit,i
               marked(j) = .true.
            end do
         end if
      end do
      primes = pack([(i, i=2,limit)], .not. marked(2:))
   end function primes

end module sieve

File: solutions/zig/sieve/sieve.zig
pub fn primes(buffer: []u32, limit: u32) []u32 {
    if (limit < 2) {
        return buffer[0..0];
    }

    var marked: [1001]bool = undefined;
    var i: u32 = 0;

    for (2..limit + 1) |n| {
        if (marked[n]) {
            continue;
        }
        for (n..limit + 1) |m| {
            if (m % n == 0) {
                marked[m] = true;
            }
        }
        buffer[i] = @intCast(n);
        i += 1;
    }
    return buffer[0..i];
}
